### ISSUE TITLE
Add 'CSV Serialization' module to GovCMS8

### ISSUE THREAD

**Issue by joshua-graham-adelphi** [govCMS/govcms8-installer#395]
Request to add Drupal module 'CSV Serialization' to GovCMS 8.

This module would be used by a number of large agencies such as DFAT's Smart Traveller GovCMS site.

Additionally a number of agencies of at least 12 are departing from Dept of Employment Shared services and are using this module in their Drupal 7 build and would like this feature included in their Drupal 8 rebuild in GovCMS.

## What value does this module/package add to GovCMS?

Ability to export Views data in CSV format so excel processing be done by admin staff.
Output is also useful for website visitors who request lists of information from the website. (e.g. FOI requests) 

## Please provide a brief outline of what this module does.

- Add CSV format to Views REST APIs.

## Who does this module benefit:
[X] end users
[X] content editors
[ ] site builders
[ ] themers
[ ] developers
 
## How could you provide/replicate the functionality of this module using alternative methods, eg in your theme?

Alternate methods:

1) JSON + free online converter:

Create a JSON views REST endpoint, download the JSON data and use a free online converter to transform to CSV.

2) Write Hack code to produce CSV file:

Basically using an existing Views JSON REST endpoint, you could add a CSV creation library to the theme layer, then use hook_views_post_render() with the extra library bloat to transform the format from JSON to CSV and change the content type HTTP header to 'text/csv'.

Example code:

```php
use Drupal\views\Render\ViewsRenderPipelineMarkup;
use Drupal\views\ViewExecutable;
use Drupal\Component\Serialization\Json;

/**
 * Implements hook_views_post_render().
 */
function mytheme_views_post_render(ViewExecutable $view, &$output, $cache) {
  if ($view->id() == 'my_csv_view') {
    // Decode the JSON.
    $json_array = Json::decode($output['#markup']);
    $csv_output = call_csv_convert_library($json_array);
    $output = [
      '#markup' => ViewsRenderPipelineMarkup::create($csv_output),
      '#cache' => $output['#cache'],
    ];
    // Change the HTTP Mime content type.
    $view->display_handler->setMimeType('text/csv');
    $view->element['#content_type'] = $view->display_handler->getMimeType();
  } 
}
```

## If this module styles or alters HTML or JavaScript output, can the functionality be provided via the theme? What alternatives have you considered.

See above comment.

## What is the maintenance and support status of the module. Describe the issue queue activity.

Module status:
- Beta version 20 september 2019
- Stable version 17 July 2018

Last commit:
 - 19 Nov 2019
 
Amount of commits:
- Over 20 committers, with one committer has 14 commits.

Issues:
- 25 open issues

Recent activity:
- 1 week ago.

## What permissions are needed to utilise the module (and are any new permissions provided by the module)?

No new permissions. Existing OOTB REST permissions are used.

## Does the module modify the database structure and/or store additional metadata on nodes or other entities? If so, why? What are the risks for future updates?

No changes to the database. (other than standard export config - serialization.settings.yml)

## Is the module designed to capture anonymous user data?

No.

## Is the output of the module typically fully cacheable? Would the inclusion of this module potentially render pages uncacheable.

Content is fully cached by Drupal Cache API.

Confirmed with below hook.

```php
use Drupal\views\Render\ViewsRenderPipelineMarkup;
use Drupal\views\ViewExecutable;
// From League\Csv required by new module.
use League\Csv\Writer;

/**
 * Implements hook_views_post_render().
 */
function govcms8_uikit_starter_views_post_render(ViewExecutable $view, &$output, $cache) {
  // Add exit function after api is called at least once.
  // Notice that that it is not called and date below does not change.
  // exit();
  if ($view->id() == 'content_rest') {
    // Load the CSV document from a string - #markup.
    $csv = Writer::createFromString(render($output));
    // Insert a new dynamic record.
    $csv->insertAll([date(DATE_RFC822)]);
    // Convert to string.
    // Calls trait League\Csv\Config->__toString().
    $csv_output = (string) $csv;
    $output = [
      '#markup' => ViewsRenderPipelineMarkup::create($csv_output),
      '#cache' => $output['#cache'],
    ];
  }
}
```

## What is your assessment of the quality of this module, the contribution history of the module's maintainers, and the uptake of the module within the Drupal community?

Module is fairly small and quality seems very good so issue queue activity is very good for small module.

## Additional information

N/A.

**Comment by simesy** [comment]
kudos for the code examples!

**Comment by isaactownsend** [comment]
Agreed this should be a standard part of the distribution - we've had other large government clients request this for reporting and content audit purposes.

**Comment by Belhendo** [comment]
Hi @joshua-graham-adelphi ,

Thank you for the request. We've discuss internally and we will need greater confidence that more agencies would need this module, it seems quite edge case to us at this stage. 

I'll leave this ticket open so others can contribute.

**Comment by catwells** [comment]
Adding further support to this request to become part of the D9 distribution. Requested enhancements to one of our departmental managed sites would use this feature and save us from writing code to do it.
Thanks
Cathy
